### ISSUE TITLE
Is it possible to set different connection settings for sync connect and async reconnect?

### ISSUE THREAD

**Issue by SX91** [hazelcast/hazelcast-cpp-client#475]
I tried to setup the sync connect / async reconnect client. But if I set ConnectionAttemptLimit to anything non-zero, it breaks async reconnection,as if it fails within ConnectionAttemptLimit tries, client just shuts down. Recreating the client is not a good option, and I couldn't find any API to restart reconnect process.

So for now I'm stuck with async connect/async reconnect with ConnectionAttemptLimit set to 0.

Is there any workaround? Or any API planned to address the issue?

**Comment by SX91** [comment]
Hello, @ihsandemir 

If I
* set ConnectionAttemptLimit to zero,
* set AsyncStart to false,
* set ReconnectMode to ASYNC

and then use the example above, client would block until the connection would be established.

I want to have sync connect (with finite number of retries) and async reconnect (with infinite number of retries). But client uses ConnectionAttemptLimit value for both cases.

**Comment by ihsandemir** [comment]
@SX91 I see your problem now. Yes, the current design is to use the same retry limit for both initial connection as well as the reconnections. Hence, there is only one configurable option. You want to separate the two cases and add something like ConnectionAttemptLimitForReconnections. Why do you need to differentiate the initial connection and reconnections in your program? If the client fails fast during initial connection, it seems to be OK for you but you do not want it to fail the connection to cluster once the client started. is this is really something you really want? 

 I will also add our product management to your request and let's discuss if it is a feature we want to add and it can definitely be added if it is a need for the users.

**Comment by ihsandemir** [comment]
@SX91 
One workaround solution:

Make start async as well. Add LifecycleListener in the config and implement the listener so that you can keep and get the state of the client at any time (see https://github.com/hazelcast/hazelcast/blob/master/hazelcast-client/src/main/java/com/hazelcast/client/util/ClientStateListener.java) and your application logic shuts down the client if a timeout passed and the initial state is not CONNECTED. Of course this logic should apply only for initial connection.

This may achieve what you want.

**Comment by dbrimley** [comment]
Would Exponential Back off work in this scenario?  

https://docs.hazelcast.org/docs/3.11/manual/html-single/index.html#configuring-client-connection-retry

We've just introduced this in the most recent release of the Java Client, we could add the same to CPP.

**Comment by SX91** [comment]
@ihsandemir 
> If the client fails fast during initial connection, it seems to be OK for you but you do not want it to fail the connection to cluster once the client started. is this is really something you really want?

Yes, this is exactly what I want. We use client to sync data across multiple instances of our server, so starting with no connection to hazelcast nodes is a failure (because of no initial data available), but losing the connection during the normal work is not.

> I will also add our product management to your request and let's discuss if it is a feature we want to add and it can definitely be added if it is a need for the users.

Thank you! Would be awesome to be able to prevent Client from shitting down with ASYNC reconnect strategy.

@dbrimley 
> Would Exponential Back off work in this scenario?

I'll look into it tomorrow, thank you!
